### PR TITLE
Saw python api

### PR DESCRIPTION
**Pull request by weaversa**

### CODE REVIEW

**Comment by WeeknightMVP** [comment]
Marking as `Draft` while changes remain underway.

Comments forthcoming...

**Comment by WeeknightMVP** [comment]
```python
from saw_client import *
```

We should avoid wildcard imports.  Import the package and use qualified names, or list each definition (with aliases to avoid any conflicts).  Some relevant guiding principles from Tim Peters's [Zen of Python](https://en.wikipedia.org/wiki/Zen_of_Python):

> Explicit is better than implicit.
> Flat is better than nested.
> Readability counts.
> Special cases aren't special enough to break the rules.
> Although practicality beats purity.
> In the face of ambiguity, refuse the temptation to guess.
> Namespaces are one honking great idea -- let's do more of those!

**Comment by WeeknightMVP** [comment]
`cry_f` and friends are mostly a wrapper around [formatted string literals](https://docs.python.org/3/tutorial/inputoutput.html#formatted-string-literals) called "f-strings".  Mentioning this will give experienced Python programmers a quick head start.

**Comment by WeeknightMVP** [comment]
Regarding `unittest` integration, saying that we _need_ to do so is a bit strong.  I'd go so far as to claim that such integration is more for developers to test the SAW Remote API than for users that call it.  (Users can integrate it with other testing frameworks or omit testing frameworks entirely.)  But it's convenient in this case.

`self.assertTrue(result.is_success())` is a more readable form.  (This assumes `is_success` returns a `bool`; otherwise `self.assertIs(result.is_success(), True)` is stronger due to "truthiness" interpretation of other types.)

**Comment by WeeknightMVP** [comment]
> SAW does not have inductive reasoning capabilities.

For SAWScript and the SAW Remote API, this is true.  But SAWCore terms can be [formalized in Coq](https://github.com/GaloisInc/saw-script/tree/master/saw-core-coq), a much more powerful theorem prover that has inductive reasoning (and many more) capabilities.  This is obviously well beyond the scope of the course; I just think this statement should be clarified.

**Comment by WeeknightMVP** [comment]
> An alternative to passing in lemmas is uninterpreting a funciton [sic]. This is useful when you don't care about the specifics of the values produced by a total function, but rather, the types of the value produced.

This is one case, yes.  It is also useful _in conjunction with overrides_, reducing proof complexity by decomposing a complex _specification_ into manageable logical steps (constrained by the overrides), much as verifying function implementations and using results as overrides does for the corresponding _implementation_.  I am trying to communicate this in the pending [`salsa20-markdown`](https://github.com/weaversa/cryptol-course/pull/190) branch for [Galois's Salsa20 example](https://saw.galois.com/intro/Salsa20.html), which naturally demonstrates the value of uninterpreted functions for the same reason it so effectively demonstrates overrides -- the implementation was derived directly from the specification.

File: saw-python/src/rcs.c [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+
+uint32_t RCS(uint32_t bits, uint32_t shift) {
+  shift %= 64;
```
Comment (weaversa) [inline saw-python/src/rcs.c:6]: That `64` should be `32`. Or, for consistency w/ the line two below, we could write `sizeof(bits)*8`.

**Comment by weaversa** [comment]
>

**Comment by weaversa** [comment]
>

**Comment by weaversa** [comment]
> `cry_f` and friends are mostly a wrapper around [formatted string literals](https://docs.python.org/3/tutorial/inputoutput.html#formatted-string-literals) called "f-strings". Mentioning this will give experienced Python programmers a quick head start.

Added a comment to this effect.

File: labs/SAW/SAW.md [hunk 2]
```diff
@@ -0,0 +1,987 @@
...
+[height-balanced binary
+trees](https://en.wikipedia.org/wiki/Self-balancing_binary_search_tree)
+on n nodes is given by the ceiling of `log(n)`. In C we might use the
+`math` library to write the helper function.
+
+```C
+#include <math.h>
+
+uint64_t log2i(uint64_t i) {
+  return ceil(log2(i));
```
Comment (weaversa) [inline labs/SAW/SAW.md:883]: @smithnormform `ceil(log2(i))` in C is not equal to Cryptol's `lg2` for values `1 << j + 1 where j >= 49`. To stick w/ a builtin function that needs an assumption, you could instead use:

```
int __builtin_clzll (unsigned long long)
```
More info here: https://en.wikipedia.org/wiki/Find_first_set#CLZ

Comment (weaversa) [reply labs/SAW/SAW.md:883]: I guess you could instead simply mention the disparity here and add a precondition to the SAW contract.

File: labs/SAW/SAW.md [hunk 3]
```diff
@@ -0,0 +1,987 @@
...
+    
+    self.execute_func(a_p, b_p)
+    
+    (c, c_p) = ptr_to_fresh(self, array_ty(5, i32), name="c")
+    self.postcondition_f("{c} == rowAdd {a} {b}")
+    
+    self.returns(c_p)
+```
+
+One could replace `self.postcondition_f("{aPost} == rowAdd {a} {b}")`
```
Comment (michaelabernardo) [inline labs/SAW/SAW.md:643]: Should this be `self.postcondition_f("{c} == rowAdd {a} {b}")`?

**Comment by djmorel** [comment]
> What is this `i*0` accomplishing?
> 
> ```
> Main> :prove \a-> initByteSequence`{5} a == repeat a
> :prove \a-> initByteSequence`{5} a == repeat a
> Q.E.D.
> (Total Elapsed Time: 0.015s, using "Z3")
> ```

It was a poor man's method of using Cryptol's loop notation to create a sequence of a specified size. The `initByteSequence` function has been replaced by Cryptol's cleaner built-in `repeat` function :)

File: labs/SAW/SAW.md [hunk 4]
```diff
@@ -1193,7 +1193,119 @@ example
 
 ## Exercise: Resolving an Attack!
 
-Feeling pretty confident with our little `player_t` and `character_t` structs? How about we go for a full on attack then? Well, an attack in our game of course between two characters ;)
+Feeling pretty confident with our little `player_t` and `character_t` structs? How about we go for a full on attack then? Well, in our game of course ;)
```
Comment (weaversa) [inline labs/SAW/SAW.md:1225]: loving the puns!

Comment (djmorel) [reply labs/SAW/SAW.md:1225]: Why thank you, it's my favorite part :)

**Comment by WeeknightMVP** [comment]
> SAW alerted us about potentially undefined behavior mentioned in the [C specification](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) has the following to say about bit shifts:

This is a phrasal portmanteau, more commonly known as a grammatical error.  Perhaps it could be rephrased to:

"SAW alerted us about potentially undefined behavior; the [C99 standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) specifies the following about bit shifts:"

([ISO 9899:2017 (C17)](https://web.archive.org/web/20181230041359if_/http://www.open-std.org/jtc1/sc22/wg14/www/abq/c17_updated_proposed_fdis.pdf) is a current specification with the same text.)

**Comment by WeeknightMVP** [comment]
> To limit code reuse we can define helper functions in Python. For example, the following construct is often used:

Code reuse is good.  Code _repetition_ is bad.  `pointer_to_fresh` avoids repetition by using an assignment to reuse the same value.  Nice and [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

**Comment by WeeknightMVP** [comment]
> SAW is telling us we forgot to add a precondition to assert our symbolic length agrees with our Python parameter self.length. This is an easy fix:

It is, but it's redundant, bad for checking (because all but one length chosen at random will ever satisfy the precondition, rendering most quickchecks vacuous), and imposes an unnecessary burden on the solver for proving.  A more direct approach is to just pass `__init__`'s `length` parameter to execute it as a concrete value:

```Python
  self.execute_func(a_p, b_p, cry(self.length))
```

**Comment by WeeknightMVP** [comment]
> We use 3 double quotes """ in our cry_f call. This technique is useful when we want to separate our expected Cryptol-defined behaviors over multiple lines to improve code readability.

Nice! This results in wide open spaces when debugging strings passed to `cry`, `cry_f`, and friends, which can be mitigated using `dedent` from the `textwrap` package that comes with Python, e.g.:

```python
from textwrap import dedent
# ...

    self.returns(cry(dedent(strip("""
      (y, z)
        where
           y = "foo"
           z = "bar"
    """))))
```

This renders (without leading/trailing whitespace) as:

```
(y, z)
  where
     y = "foo"
     z = "bar"
```

**Comment by WeeknightMVP** [comment]
>     character       = self.alloc(alias_ty("struct.character_t"))

This is probably supposed to be `character_p`.

**Comment by WeeknightMVP** [comment]
> So maybe we tried being too fancy showing off our associative property knowledge.

One of those arithmetic properties, anyway...

**Comment by WeeknightMVP** [comment]
> That is, was our assumption a fair one to make? If not, change the C code (possibly using different library functions) to match lg2 in Cryptol.

...and verify it!

File: labs/SAW/SAW.md (hunk 2 above)
Comment (djmorel) [reply labs/SAW/SAW.md:1861]: Updated in SAW.md now!

**Comment by djmorel** [comment]
> > SAW alerted us about potentially undefined behavior mentioned in the [C specification](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) has the following to say about bit shifts:
> 
> This is a phrasal portmanteau, more commonly known as a grammatical error. Perhaps it could be rephrased to:
> 
> "SAW alerted us about potentially undefined behavior; the [C99 standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) specifies the following about bit shifts:"
> 
> ([ISO 9899:2017 (C17)](https://web.archive.org/web/20181230041359if_/http://www.open-std.org/jtc1/sc22/wg14/www/abq/c17_updated_proposed_fdis.pdf) is a current specification with the same text.)

Updated SAW.md!

**Comment by djmorel** [comment]
> > SAW is telling us we forgot to add a precondition to assert our symbolic `length` agrees with our Python parameter `self.length`. This is an easy fix:
> 
> It is, but it's redundant, bad for checking (because all but one length chosen at random will ever satisfy the precondition, rendering most quickchecks vacuous), and imposes an unnecessary burden on the solver for proving. A more direct approach is to just execute `self.length` as a concrete value:
> 
> ```python
>   self.execute_func(a_p, b_p, cry(self.length))
> ```

Updated SAW.md!

**Comment by djmorel** [comment]
> 

Updated SAW.md!

**Pull request merged**